lib/benchee/benchmark/runner.ex

```diff
@@ -120,33 +120,52 @@ defmodule Benchee.Benchmark.Runner do
   end
   defp do_benchmark(scenario = %Scenario{run_times: run_times},
                     scenario_context) do
-    run_time = measure_call(scenario, scenario_context)
+    run_time = measure_iteration(scenario, scenario_context)
     updated_scenario = %Scenario{scenario | run_times: [run_time | run_times]}
     updated_context =
       %ScenarioContext{scenario_context | current_time: current_time()}
     do_benchmark(updated_scenario, updated_context)
   end
 
-  defp measure_call(scenario, scenario_context = %ScenarioContext{
-                      num_iterations: num_iterations
-                    }) do
-    measure_call_n_times(scenario, scenario_context) / num_iterations
+  defp measure_iteration(scenario, scenario_context = %ScenarioContext{
+                          num_iterations: num_iterations
+                        }) do
+    run_iteration(scenario, scenario_context) / num_iterations
+  end
+
+  defp run_iteration(scenario, scenario_context) do
+    run_before_each(scenario, scenario_context)
+    measure_call(scenario, scenario_context)
+  end
+
+  defp run_before_each(%{
+                          before_each: local_before_each
+                        },
+                        %{
+                          config: %{before_each: global_before_each}
+                        }) do
+    if global_before_each, do: global_before_each.()
+    if local_before_each,  do: local_before_each.()
   end
 
   @no_input Benchmark.no_input()
-  defp measure_call_n_times(%Scenario{function: function, input: @no_input},
-                            %ScenarioContext{num_iterations: num_iterations}) do
-    {microseconds, _return_value} = :timer.tc fn ->
-      RepeatN.repeat_n(function, num_iterations)
-    end
+  defp measure_call(%Scenario{function: function, input: @no_input},
+                    %ScenarioContext{num_iterations: iterations}) do
+    measure_call function, iterations
+  end
+  defp measure_call(%Scenario{function: function, input: input},
+                    %ScenarioContext{num_iterations: iterations}) do
+    fun_with_input = fn -> function.(input) end
+    measure_call fun_with_input, iterations
+  end
 
+  defp measure_call(function, 1) do
+    {microseconds, _return_value} = :timer.tc function
     microseconds
   end
-  defp measure_call_n_times(%Scenario{function: function, input: input},
-                            %ScenarioContext{num_iterations: num_iterations}) do
-    fun_with_input = fn -> function.(input) end
+  defp measure_call(function, iterations) do
     {microseconds, _return_value} = :timer.tc fn ->
-      RepeatN.repeat_n(fun_with_input, num_iterations)
+      RepeatN.repeat_n(function, iterations)
```

Comment:
Here's what I'm thinking - we might need to change this `RepeatN.repeat_n/2` function. If we move `:timer.tc/1` _inside_ that function, it can still return the `{microseconds, _return_value}` tuple, but we'll be able to control around which function we're measuring. We could then pass in the before & after functions as well.

What do you think? It seems like it wouldn't be so tricky to move that `:timer.tc/1` call in there, right?

Reply:
Hm I still don't believe that'll work. Unless I'm missing something big.

The reason we are repeating at all is because a single invocation of the function is too fast. It finishes in under a microsecond or 4 microseconds or something so that I won't trust its measurement (coming out as 3 or 4 microseconds is a huge difference there). So measuring multiple small wrong invocations (let's say they mostly come back with one microsecond) and then averaging them seems very error prone to me. Therefore what I used to do is run the small function a bunch of times (10 / 100) and then average it. I.e. if the function really took 1.3 microseconds I'd hopefully end up with 13 microseconds and report an average of 1.3 microseconds. If all the measured run times of 1 microseconds were just added up it'd be 10 microseconds and we'd be at an average of 1 microsecond which seems unsatisfactory.

That's why I don't believe we can safely measure individual invocation - doing that we could just take away the whole repeat_n functionality because then it'd just be normal benchee behaviour.

You are right though, we can't just not run the before_each so that overhead will be added but I think we should then make a warning that it's a bad idea to mix super fast functions and before_each for this reason :)

Reply:
I'm ok with a warning at first. That seems like a reasonable compromise, and one I don't think would end up being triggered very much. Most likely if folks are using some sort of `before_each` function it's to set up state on which their scenario is going to operate, and that's not going to be measured in microseconds.

If we really wanted to I bet we could get this 100% "correct", but after thinking about it further, I'm not 100% sure this will even be something users ever run into. Let's put it off until later 🕦